pkg/controllers/multiclusterservice/endpointslice_dispatch_controller.go

```diff
@@ -324,62 +320,26 @@ func (c *EndpointsliceDispatchController) dispatchEndpointSlice(ctx context.Cont
-		if err := helper.ConvertToTypedObject(unstructuredObj, endpointSlice); err != nil {
-			klog.Errorf("Failed to convert unstructured object to typed object, error is: %v", err)
+		clusterObj, err := util.GetCluster(c.Client, clusterName)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				c.EventRecorder.Eventf(mcs, corev1.EventTypeWarning, events.EventReasonClusterNotFound, "Consumer cluster %s is not found", clusterName)
+				continue
+			}
+			klog.Errorf("Failed to get cluster %s, error is: %v", clusterName, err)
 			return err
 		}
-
-		// Use this name to avoid naming conflicts and locate the EPS source cluster.
-		endpointSlice.Name = epsSourceCluster + "-" + endpointSlice.Name
-		clusterNamespace := names.GenerateExecutionSpaceName(clusterName)
-		endpointSlice.Labels = map[string]string{
-			discoveryv1.LabelServiceName:    mcs.Name,
-			workv1alpha1.WorkNamespaceLabel: clusterNamespace,
-			workv1alpha1.WorkNameLabel:      work.Name,
-			util.ManagedByKarmadaLabel:      util.ManagedByKarmadaLabelValue,
-			discoveryv1.LabelManagedBy:      util.EndpointSliceDispatchControllerLabelValue,
+		if !util.IsClusterReady(&clusterObj.Status) {
+			c.EventRecorder.Eventf(mcs, corev1.EventTypeWarning, events.EventReasonSyncServiceFailed,
+				"Consumer cluster %s is not ready, skip to propagate MultiClusterService", clusterName)
+			continue
 		}
-		endpointSlice.Annotations = map[string]string{
-			// This annotation is used to identify the source cluster of EndpointSlice and whether the eps are the newest version
-			util.EndpointSliceProvisionClusterAnnotation: epsSourceCluster,
+		if !helper.IsAPIEnabled(clusterObj.Status.APIEnablements, util.EndpointSliceGVK.GroupVersion().String(), util.EndpointSliceGVK.Kind) {
+			c.EventRecorder.Eventf(mcs, corev1.EventTypeWarning, events.EventReasonAPIIncompatible, "Consumer cluster %s does not support EndpointSlice", clusterName)
+			continue
 		}
 
-		workMeta := metav1.ObjectMeta{
-			Name:       work.Name,
-			Namespace:  clusterNamespace,
-			Finalizers: []string{util.ExecutionControllerFinalizer},
-			Annotations: map[string]string{
-				util.EndpointSliceProvisionClusterAnnotation: epsSourceCluster,
-			},
-			Labels: map[string]string{
-				util.ManagedByKarmadaLabel:             util.ManagedByKarmadaLabelValue,
-				util.MultiClusterServiceNameLabel:      mcs.Name,
-				util.MultiClusterServiceNamespaceLabel: mcs.Namespace,
-			},
-		}
-		unstructuredEPS, err := helper.ToUnstructured(endpointSlice)
-		if err != nil {
-			klog.Errorf("Failed to convert typed object to unstructured object, error is: %v", err)
-			return err
-		}
-		if err := helper.CreateOrUpdateWork(c.Client, workMeta, unstructuredEPS); err != nil {
-			klog.Errorf("Failed to dispatch EndpointSlice %s/%s from %s to cluster %s:%v",
-				work.GetNamespace(), work.GetName(), epsSourceCluster, clusterName, err)
+		if err := c.ensureEndpointSliceWork(mcs, work, epsSourceCluster, clusterName); err != nil {
 			return err
```

Comment:
Consider handling resources as much as possible.

Reply:
This PR should only focus on event, let's do it in another PR.